src/main/java/org/apache/mesos/hdfs/state/PersistentState.java

```diff
@@ -42,28 +56,168 @@ public void setFrameworkId(FrameworkID frameworkId) throws InterruptedException,
+        deadJournalHosts.add(journalHost);
+      }
+    }
+    return deadJournalHosts;
+  }
+  public List<String> getDeadNameNodes() {
+    HashMap<String, String> nameNodes = getNameNodes();
+    Set<String> nameHosts = nameNodes.keySet();
+    List<String> deadNameHosts = new ArrayList<>();
+
+    for (String nameHost : nameHosts) {
+      if (nameNodes.get(nameHost) == null) {
+        deadNameHosts.add(nameHost);
+      }
+    }
+    return deadNameHosts;
   }
+  public List<String> getDeadDataNodes() {
+    HashMap<String, String> dataNodes = getDataNodes();
+    Set<String> dataHosts = dataNodes.keySet();
+    List<String> deadDataHosts = new ArrayList<>();
 
-  public void setNamenode(Node node) throws InterruptedException, ExecutionException, IOException {
-    set(NAMENODE_KEY, node);
+    for (String dataHost : dataHosts) {
+      if (dataNodes.get(dataHost) == null) {
+        deadDataHosts.add(dataHost);
+      }
+    }
+    return deadDataHosts;
+  }
+  // TODO (nicgrayson) add tests with in memory zookeeper
+  public HashMap<String, String> getJournalNodes() {
+    return getHashMap(JOURNALNODES_KEY);
   }
 
-  public Set<Node> getNodes() throws InterruptedException, ExecutionException, IOException,
-      ClassNotFoundException {
-    Set<Node> nodes = get(NODES_KEY);
-    if (nodes == null)
-      nodes = new HashSet<Node>();
-    return nodes;
+  public HashMap<String, String> getNameNodes() {
+    return getHashMap(NAMENODES_KEY);
   }
 
-  public void setNodes(Set<Node> nodes) throws InterruptedException, ExecutionException,
-      IOException {
-    set(NODES_KEY, nodes);
+  public HashMap<String, String> getDataNodes() {
+    return getHashMap(DATANODES_KEY);
+  }
+
+  public Collection<String> getAllTaskIds() {
+    HashMap<String, String> allTasksIds = getJournalNodes();
+    allTasksIds.putAll(getNameNodes());
+    allTasksIds.putAll(getDataNodes());
+    return allTasksIds.values();
+  }
+
+  public void addHdfsNode(Protos.TaskID taskId, String hostname, String taskName) {
+    switch (taskName) {
+      case HDFSConstants.NAME_NODE_ID :
+        HashMap<String, String> nameNodes = getNameNodes();
+        nameNodes.put(hostname, taskId.getValue());
+        System.out.println("Saving the name node " + hostname + " " + taskId.getValue());
+        setNameNodes(nameNodes);
+        break;
+      case HDFSConstants.JOURNAL_NODE_ID :
+        HashMap<String, String> journalNodes = getJournalNodes();
+        journalNodes.put(hostname, taskId.getValue());
+        setJournalNodes(journalNodes);
+        break;
+      case HDFSConstants.DATA_NODE_ID :
+        HashMap<String, String> dataNodes = getDataNodes();
+        dataNodes.put(hostname, taskId.getValue());
+        setDataNodes(dataNodes);
+        break;
+      case HDFSConstants.ZKFC_NODE_ID :
+        break;
+      default :
+        log.error("Task name unknown");
+    }
+  }
+
+  public void removeTaskId(String taskId) {
```

Comment:
True, I added a TODO to optimize these methods. Another possibility is using the task id for the key which I mentioned in the TODO.